### ISSUE TITLE
Purchases not being acknowledged sometimes

### ISSUE THREAD

**Issue by LiviuSebastian** [moisoni97/google-inapp-billing#70]
Hello, and thank you for developing this library.

I have implemented version 1.1.1 of the library in two of my apps for in-app purchases. In these apps, I am also using the com.android.billingclient:billing:6.0.0 dependency. However, I am facing an issue where purchases are automatically refunded exactly after 3 days and 10 minutes.

Upon researching the Google documentation, I found that "If you don't acknowledge a purchase within three days, the user automatically receives a refund, and Google Play revokes the purchase" (source: [Google Play Billing Documentation](https://developer.android.com/google/play/billing/integrate#java)).

This automatic refund is not happening for every purchase, and I'm unsure how to resolve it. Do you have any ideas or suggestions on what steps I should take to address this issue?

Thank you for your assistance.

**Comment by newGenMobile** [comment]
Always a good idea to have redundancy. if the auto acknowledge or consume is not working. Here is part of my code which i hope helps you out.

`
   private final String ESKU = "1";
    private final String ESKU_SUB = "2";
    private final String ESKU_SUB_DEAL = "3";
    private final String SmallDonation = "small";
    private final String MediumDonation = "medium";
    private final String LargeDonation = "large";

@Override
            public void onProductsPurchased(@NonNull List<PurchaseInfo> purchases) {
                String product;
                String purchaseToken;

                for (PurchaseInfo purchaseInfo : purchases) {
                    product = purchaseInfo.getProduct();
                    purchaseToken = purchaseInfo.getPurchaseToken();

                    Utils.logError("BillingConnector", "Product purchased: " + product);
                    Utils.logError("BillingConnector", "Purchase token: " + purchaseToken);

                    purchasedInfoList.add(purchaseInfo); //check "usefulPublicMethods" to see how to acknowledge or consume a purchase manually

                    // Manual acknowledgment or consumption as redundancy
                    checkAndAcknowledgePurchase(purchaseInfo);
                }
            }

private void checkAndAcknowledgePurchase(PurchaseInfo purchaseInfo) {
        if (purchaseInfo != null && billingConnector != null && billingConnector.isReady()) {
            if (isAcknowledgableSKU(purchaseInfo.getProduct())) {
                if (!purchaseInfo.isAcknowledged()) {
                    billingConnector.acknowledgePurchase(purchaseInfo);
                    Utils.logError("BillingManager", "checkAndAcknowledgePurchase: Manual acknowledgePurchase");
                }
            } else if (isConsumableSKU(purchaseInfo.getProduct())) {
                billingConnector.consumePurchase(purchaseInfo);
                Utils.logError("BillingManager", "checkAndAcknowledgePurchase: Manual consumePurchase");
            }
        } else {
            Utils.logError("BillingManager", "checkAndAcknowledgePurchase: Billing connector is not ready.");
        }
    }

    private boolean isConsumableSKU(String product) {
        // Check if the purchased product is one of the consumable SKUs
        return SmallDonation.equals(product) || MediumDonation.equals(product) || LargeDonation.equals(product);
    }

    private boolean isAcknowledgableSKU(String product) {
        // Check if the purchased product is one of the acknowledgeable SKUs
        return ESKU.equals(product) || ESKU_SUB.equals(product) || ESKU_SUB_DEAL.equals(product);
    }`

**Comment by Touchsoftwares** [comment]
billingConnector = new BillingConnector(this, getString(R.string.license_key))
                .setConsumableIds(listConsumables)
                .setNonConsumableIds(listProducts)
                .setSubscriptionIds(listSucriptions)
                .autoAcknowledge() <------
               // .autoConsume()
                .enableLogging() 
                .connect();

**Comment by LiviuSebastian** [comment]
This is my code. Although I have "autoAcknowledge()" still some times after after 3 days and 10 minutes the purchase is refunded:

billingConnector = new BillingConnector(this, getString(R.string.license_key))
                    .setNonConsumableIds(nonConsumableId)
                    .autoAcknowledge()
                    .enableLogging()
                    .connect();

**Comment by duckyngo** [comment]
I got exactly same problem. Here is my code for initialize BillingClient:
```
billingConnector = BillingConnector(
            requireContext(),
            BASE_64_ENCODED_PUBLIC_KEY) //"license_key" - public developer key from Play Console
//            .setConsumableIds(consumableIds) //to set consumable ids - call only for consumable products
            .setNonConsumableIds(nonConsumableIds) //to set non-consumable ids - call only for non-consumable products
            .setSubscriptionIds(subscriptionIds) //to set subscription ids - call only for subscription products
            .autoAcknowledge() //legacy option - better call this. Alternatively purchases can be acknowledge via public method "acknowledgePurchase(PurchaseInfo purchaseInfo)"
            .autoConsume() //legacy option - better call this. Alternatively purchases can be consumed via public method consumePurchase(PurchaseInfo purchaseInfo)"
            .enableLogging() //to enable logging for debugging throughout the library - this can be skipped
            .connect() //to connect billing client with Play Console
```

Do you have any ideas or suggestion to address this issue?

**Comment by akardas16** [comment]
actually if you reinitialize the billingConnector, library will acknowledge all non-acknowledged packages.

**Comment by akardas16** [comment]
I found that if you try to re-acknowledge an acknowledged purchase, `Server error! Please try again` error will occur in  `billingClient.acknowledgePurchase(acknowledgePurchaseParams) { billingResult: BillingResult -> }` so if you get error about acknowledge, probably purchase already acknowledged.

**Comment by moisoni97** [comment]
Sometimes the Billing API can't acknowledge the purchase right away, so you need to initialize the billing client later.
Use error code: 'ACKNOWLEDGE_WARNING' to inform the user to come back later to receive the purchase.
This is how it currently works.

**Comment by akardas16** [comment]
> [@LiviuSebastian](https://github.com/LiviuSebastian) [@duckyngo](https://github.com/duckyngo) [@akardas16](https://github.com/akardas16) [@moisoni97](https://github.com/moisoni97) [@Touchsoftwares](https://github.com/Touchsoftwares) Have it been fixed with the latest version 1.1.5 ?

I dont think this is related to the library. if you enable autoacknowledge this shouldn't be an issue because if user reopen the app, library will acknowledge purchases but you have to initiate billing client when app opened otherwise user may not check again purchase screen if you initiate billing client in purchase screen
